src/language-mode.coffee

```diff
@@ -255,12 +255,32 @@ class LanguageMode
       precedingRow = bufferRow - 1
       return currentIndentLevel if precedingRow < 0
 
-    precedingLine = @buffer.lineForRow(precedingRow)
     desiredIndentLevel = @editor.indentationForBufferRow(precedingRow)
-    desiredIndentLevel += 1 if increaseIndentRegex.testSync(precedingLine) and not @editor.isBufferRowCommented(precedingRow)
+    unless @editor.isBufferRowCommented(precedingRow)
+      precedingLine = @buffer.lineForRow(precedingRow)
+      desiredIndentLevel += 1 if increaseNextIndentRegex.testSync(precedingLine)
+
+    decreaseNextIndentRegex = @decreaseNextIndentRegexForScopeDescriptor(scopeDescriptor)
+    decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
+    return desiredIndentLevel unless decreaseNextIndentRegex or decreaseIndentRegex
+
+    if decreaseNextIndentRegex?.testSync(precedingLine) or (decreaseIndentRegex?.testSync(tokenizedLine.text) and not @editor.isBufferRowCommented(bufferRow))
+      lastIndentingRow = bufferRow
+      while (lastIndentingRow = @buffer.previousNonBlankRow(lastIndentingRow))?
+        continue if @editor.isBufferRowCommented(lastIndentingRow)
 
-    return desiredIndentLevel unless decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
-    desiredIndentLevel -= 1 if decreaseIndentRegex.testSync(tokenizedLine.text)
+        lastIndentingLine = @buffer.lineForRow(lastIndentingRow)
+        if increaseNextIndentRegex.testSync(lastIndentingLine)
+          desiredIndentLevel = @editor.indentationForBufferRow(lastIndentingRow) + 1
+
+          for row in [lastIndentingRow..bufferRow]
+            continue if @editor.isBufferRowCommented(row)
+
+            rowLine = @buffer.lineForRow(row)
+            desiredIndentLevel -= 1 if decreaseNextIndentRegex?.testSync(rowLine)
+            desiredIndentLevel -= 1 if decreaseIndentRegex?.testSync(rowLine)
+
+          return Math.max(desiredIndentLevel, 0)
```

Comment:
Of course... To determine the correct level of indentation it will search for the most recent line that was indented and then calculate the current indent level based on the rows between the last indented row and the current row.

So the first part of the while loop searches for a previous line that matches the indent pattern. As soon as 1 is found, it will enter the for loop and process every row between the two. After entering the for loop, it will never get back to the while loop but will return the function there.

Does this help or do you want/need more insights/reasoning?